future/atomicTransactionComposer.go

```diff
@@ -0,0 +1,540 @@
+		onComplete,
+		nil,
+		nil,
+		types.StateSchema{},
+		types.StateSchema{},
+		suggestedParams,
+		sender,
+		note,
+		types.Digest{},
+		lease,
+		rekeyTo)
+
+	if err != nil {
+		return err
+	}
+
+	txAndSigner := TransactionWithSigner{
+		Txn:    tx,
+		Signer: signer,
+	}
+
+	atc.transactions = append(atc.transactions, txAndSigner)
+	atc.methodCalls = append(atc.methodCalls, method)
+	return nil
+}
+
+/**
+* Finalize the transaction group and returned the finalized transactions.
+*
+* The composer's status will be at least BUILT after executing this method.
+ */
+func (atc *AtomicTransactionComposer) BuildGroup() ([]TransactionWithSigner, error) {
+	if atc.status > BUILDING {
+		return atc.transactions, nil
+	}
+
+	var txns []types.Transaction
+	for _, txAndSigner := range atc.transactions {
+		txns = append(txns, txAndSigner.Txn)
+	}
+
+	gid, err := crypto.ComputeGroupID(txns)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, txAndSigner := range atc.transactions {
+		txAndSigner.Txn.Group = gid
+	}
+
+	atc.status = BUILT
+	return atc.transactions, nil
+}
+
+/**
+* Obtain signatures for each transaction in this group. If signatures have already been obtained,
+* this method will return cached versions of the signatures.
+*
+* The composer's status will be at least SIGNED after executing this method.
+*
+* An error will be thrown if signing any of the transactions fails.
+*
+* @returns A promise that resolves to an array of signed transactions.
+ */
+func (atc *AtomicTransactionComposer) GatherSignatures() ([][]byte, error) {
+	// if status is at least signed then return cached signed transactions
+	if atc.status >= SIGNED {
+		return atc.signedTxs, nil
+	}
+
+	// retrieve built transactions and verify status is BUILT
+	txs, err := atc.BuildGroup()
+	if err != nil {
+		return nil, err
+	}
+
+	var sigTxs [][]byte
+	var txids []string
+	for _, txAndSigner := range txs {
+		txGroup := []types.Transaction{txAndSigner.Txn}
+		indexesToSign := []int{0}
+		sigStx, txid, err := txAndSigner.Signer(txGroup, indexesToSign)
+		if err != nil {
+			return nil, err
+		}
+
+		sigTxs = append(sigTxs, sigStx[0])
+		txids = append(txids, txid[0])
+	}
+
+	atc.signedTxs = sigTxs
+	atc.txids = txids
+	atc.status = SIGNED
+	return sigTxs, nil
+}
+
+/**
+* Send the transaction group to the network, but don't wait for it to be committed to a block. An
+* error will be thrown if submission fails.
+*
+* The composer's status must be SUBMITTED or lower before calling this method. If submission is
+* successful, this composer's status will update to SUBMITTED.
+*
+* Note: a group can only be submitted again if it fails.
+*
+* @returns A promise that, upon success, resolves to a list of TxIDs of the submitted transactions.
+ */
+func (atc *AtomicTransactionComposer) Submit(client *algod.Client) ([]string, error) {
+	if atc.status > SUBMITTED {
+		return nil, errors.New("status must be SUBMITTED or lower in order to call Submit()")
+	}
+
+	if atc.status == SUBMITTED {
```

Comment:
This should not be here, so that resubmission is allowed if the transactions were not committed.